libc/stat.c

```diff
@@ -0,0 +1,503 @@
+	TEST_ASSERT_EQUAL_INT(new_time, buffer1.st_mtim.tv_sec);
+
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer1));
+	TEST_ASSERT_EQUAL_INT(buffer1.st_ctim.tv_sec, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_atim.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(new_time, buffer1.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_mtim.tv_sec, buffer.st_mtim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(new_time, buffer1.st_mtim.tv_sec);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer1));
+	TEST_ASSERT_EQUAL_INT(buffer1.st_ctim.tv_sec, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_atim.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(new_time, buffer1.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_mtim.tv_sec, buffer.st_mtim.tv_sec);
+	TEST_ASSERT_EQUAL_INT(new_time, buffer1.st_mtim.tv_sec);
+#endif
+}
+
+////////////////////////////////////
+
+TEST_SETUP(stat_errno)
+{
+	fd = open(path, O_CREAT | O_RDONLY, 0666);
+}
+
+
+TEST_TEAR_DOWN(stat_errno)
+{
+	close(fd);
+	remove(path);
+}
+
+
+TEST(stat_errno, eacces)
+{
+	int errno_fd;
+	const char *pathErrno = "/root/somefile.txt";
+	struct stat buffer;
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat(pathErrno, &buffer));
+	TEST_ASSERT_TRUE(errno == ENOENT || EACCES);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, lstat(pathErrno, &buffer));
+	TEST_ASSERT_TRUE(errno == ENOENT || EACCES);
+
+	errno = 0;
+	errno_fd = open(pathErrno, O_RDWR);
+	TEST_ASSERT_EQUAL_INT(-1, fstat(errno_fd, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	close(errno_fd);
+}
+
+
+TEST(stat_errno, ebadf)
+{
+	struct stat buffer;
+	int fd_invaild = open("never_existed.txt", O_RDONLY);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(INT_MAX, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(-1, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(fd_invaild, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+}
+
+
+TEST(stat_errno, eloop)
+{
+	int sym_fd;
+	struct stat buffer;
+
+	unlink(symPath);
+	unlink("sec_symlink");
+
+	TEST_ASSERT_EQUAL_INT(0, symlink("sec_symlink", symPath));
+	TEST_ASSERT_EQUAL_INT(0, symlink(symPath, "sec_symlink"));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat(symPath, &buffer));
+	TEST_ASSERT_EQUAL_INT(ELOOP, errno);
+
+	errno = 0;
+	sym_fd = open(symPath, O_RDONLY);
+	TEST_ASSERT_EQUAL_INT(-1, fstat(sym_fd, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	close(sym_fd);
+
+	TEST_ASSERT_EQUAL_INT(0, unlink(symPath));
+	TEST_ASSERT_EQUAL_INT(0, unlink("sec_symlink"));
+}
+
+
+TEST(stat_errno, enametoolong)
+{
+	struct stat buffer;
+	const char *errnoPath = "LoremipsumdolorsitametconsecteturadipiscingelitCrasdictumnislquisgravidasemperEtiamfringillaliberorisusidcongueipsumvariusatCrassitametsapiennecpurusplacerateleifendasitametnislNamdictumlacussitametvelitmorbiLoremipsumdolorsitametconsecteturadipiscingelitCrasdictumnislquisgravidasemperEtiamfrin.txt";
```

Comment:
I'd rather name it:

```suggestion
	const char *tooLongPath = "LoremipsumdolorsitametconsecteturadipiscingelitCrasdictumnislquisgravidasemperEtiamfringillaliberorisusidcongueipsumvariusatCrassitametsapiennecpurusplacerateleifendasitametnislNamdictumlacussitametvelitmorbiLoremipsumdolorsitametconsecteturadipiscingelitCrasdictumnislquisgravidasemperEtiamfrin.txt";
```

or sth similar